### PR TITLE
Lazy loading + prefetching

### PR DESCRIPTION
**Pull request by Velenir**
- Lazy load all the things
- Add `<link rel="prefetch">` for all the things
- Make lots of sync code into async
- Dynamic wrapper for `react-toastify` that inits on first call
- Add a placeholder for `EtherscanLink` to avoid layout jitter

Some stuff may be broken as a result, though I tested

Screenshot from 2020-02-07 14-50-27

Audit json:

[localhost_5000-20200207T132733.zip](https://github.com/gnosis/dex-react/files/4170642/localhost_5000-20200207T132733.zip)

Possible further improvements:
- Host Token images locally and optimize them
- Images from github are not very web-friendly

### CODE REVIEW

**Comment by gnosis-info** [comment]
Travis automatic deployment:
 https://pr502--dexreact.review.gnosisdev.com

File: src/components/EtherscanLink.tsx [hunk 1]
```diff
@@ -29,8 +30,14 @@ function getEtherscanDomainSuffix(type: EtherscanLinkType, identifier: string):
...
-export const EtherscanLink: React.FC<EtherscanLinkProps> = ({ type, identifier, label, className }) => {
-  const { networkId } = useWalletConnection()
+export const EtherscanLink: React.FC<EtherscanLinkProps> = ({
+  type,
+  identifier,
+  label,
+  className,
+  networkId: netID,
+}) => {
+  const { networkId = netID } = useWalletConnection()
```
Comment (alfetopito) [inline src/components/EtherscanLink.tsx]: Not for this PR, but maybe we should default to mainnet when wallet not connected?
I'd like to see the verified main contract even when I'm not connected.

File: src/toastify.ts [hunk 2]
```diff
@@ -0,0 +1,43 @@
+import { Toast } from 'react-toastify'
```
Comment (anxolin) [inline src/toastify.ts]: Why did you need to create this component?

I've got a bit lost in the code below. I get that you override/wrap the methods of toast, probably trying to lazy load it.

My understanding is you only load it's css and the `toast` export (I understand that's the biggest part of it), when you use one of it's methods. Also you delay the configuration of it.

am I getting it somewhat right?

File: src/App.tsx [hunk 3]
```diff
@@ -70,22 +85,23 @@ const App: React.FC = () => (
...
-          <PrivateRoute path="/liquidity" exact component={Strategies} />
-          <PrivateRoute path="/wallet" exact component={Wallet} />
-          <PrivateRoute path="/orders" exact component={Orders} />
-          <Route path="/about" exact component={About} />
-          <Route path="/source-code" exact component={SourceCode} />
-          <Route path="/connect-wallet" exact component={ConnectWallet} />
-          <Redirect from="/" to="/trade/DAI-USDC?sell=0&buy=0" />
-          <Route component={NotFound} />
-        </Switch>
+        <React.Suspense fallback={null}>
```
[the remark this answers is not in the document]
Comment (Velenir) [reply src/App.tsx]: There generally won't be an error as it loads what is always available -- chunk files
Plus that's how Suspense works -- it catches a thrown Promise and shows fallback, when promise resolves -- shows actual component

But it may be a good idea to add a general ErrorBoundary above Suspense
Sometime later

File: src/App.tsx [hunk 4]
```diff
@@ -70,22 +85,23 @@ const App: React.FC = () => (
...
+            <Route path="/connect-wallet" exact component={ConnectWallet} />
+            <Redirect from="/" to="/trade/DAI-USDC?sell=0&buy=0" />
+            <Route component={NotFound} />
+          </Switch>
+        </React.Suspense>
       </Layout>
     </Router>
     {process.env.NODE_ENV === 'development' &&
-      Web3Connect.isMobile() &&
+      require('web3connect').default.isMobile() &&
```
[the remark this answers is not in the document]
Comment (Velenir) [reply src/App.tsx]: No need to lazy load as we don't load at all in production.
As for moving the whole condition to `Console` component, I remember it made Web3Connect chunk  sync loaded for me previously, but I tried now and all is good for some reason.
Will change.

File: src/PrivateRoute.tsx [hunk 5]
```diff
@@ -0,0 +1,59 @@
+import React, { useEffect } from 'react'
+import { Route, RouteProps, Redirect } from 'react-router-dom'
+import useSafeState from 'hooks/useSafeState'
+
+const useIsWalletConnected = (): {
```
[the remark this answers is not in the document]
Comment (Velenir) [reply src/PrivateRoute.tsx]: I opted not to move it to hooks as that one was made specifically for PrivateRoute to facilitate correct code splitting boundaries and shouldn't be used anywhere else.
So I'd prefer it confined to this file.

File: src/App.tsx [hunk 6]
```diff
@@ -2,64 +2,79 @@ import 'types'
...
+  import(
+    /* webpackChunkName: "Extra_routes_chunk"*/
+
+    'pages/About'
+  ),
+)
+const Trade = React.lazy(() =>
+  import(
+    /* webpackChunkName: "Trade_chunk"*/
+
```
[the remark this answers is not in the document]
Comment (Velenir) [reply src/App.tsx]: I don't know how it got there :crab:

File: src/toastify.ts (hunk 2 above)
Comment (Velenir) [reply src/toastify.ts]: Yes
So `setupToastify.ts` which includes the library and css synchronously is like that to be put in one chunk
Then in `toastify.ts` it is imported dynamically and reexports a sort of proxy that wraps every method and makes it a promise. That way the interface stays the same, but every method is promisified. `setupToastify` code is run on the first invocation of `toast.<any_method>`, then cached thanks to how dynamic import works.
Thus where we actually need to use it we have:
```
import { toast } from 'toastify'
// ...
toast.success(`The token ${tokenDisplayName} has been enabled for trading`)
```


An alternative would be:
```ts
import('setupToastify').then(({toast}) => toast.success(`The token ${tokenDisplayName} has been enabled for trading`))
```
every time we would need a toast :clinking_glasses:
